src/pytorch_lightning/core/saving.py

```diff
@@ -250,7 +256,8 @@ def _load_state(
     obj = cls(**_cls_kwargs)
 
     # give model a chance to load something
-    obj.on_load_checkpoint(checkpoint)
+    if not isinstance(obj, ModelIO):
```

Comment:
added to block mypy from raising union-attr error when obj is ModelIO

Reply:
I can't see that error locally. Could we revert it and see if it works?
It shouldn't be necessary because you corrected the annotation of cls above, and hence mypy should be able to infer that the obj should be of that type.

Reply:
updated in https://github.com/Lightning-AI/lightning/pull/13932/commits/855679fa49c7f0b18e1882008483377aee6efd71